server.py

```diff
@@ -38,14 +38,36 @@ def index():
 def list_products():
     category        = request.args.get('category')
     discontinued    = str2bool(request.args.get('discontinued'))
-    results     = []
+    min_price       = request.args.get('min-price')
+    max_price       = request.args.get('max-price')
+    price           = request.args.get('price')
+    limit           = request.args.get('limit')
+    results         = []
+    count           = 0
+    cutoff          = 2
+
+    if (limit is not None and int(limit) >= 0):
+        cutoff = int(limit)
+    elif (limit is not None and int(limit) < 0):
+        return reply(results, HTTP_400_BAD_REQUEST)
+
     for key, value in products.iteritems():
-        if matchesClause(category,      value['category']) and \
-           matchesClause(discontinued,  value['discontinued']):
-             results.append(products[key])
+        if (count == cutoff):
+            break
+        if matches_clause(category,      value['category']) and \
+           matches_clause(discontinued,  value['discontinued']) and \
+           matches_price(min_price, max_price, price, int(value['price'])):
```

Comment:
As mentioned below, it's probably better to store 'price' into its desired data-type (int or decimal) when it is loaded into the program.

That way, we need not have to handle the casting of the data-type here, e.g. `int(value['price'])`